### PR TITLE
[Naitei] Write_Controller_Product_API

### PR DESCRIPTION
**Pull request by ThaiKhacDuoc**
## Related Tickets
- ticket redmine

## WHAT (optional)
- Change number items `completed/total` in admin page.

## HOW
- I edit js file, inject not_vary_normal items in calculate function.

## WHY (optional)
- Because in previous version - number just depends on `normal` items. But in new version, we have `state` and `confirm_state` depends on both `normal` + `not_normal` items.

### CODE REVIEW

**Comment by ThaiKhacDuoc** [comment]
ready

File: app/controllers/api/v1/admin/products_controller.rb [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    render json: {
+      status: true,
+      data: ActiveModelSerializers::SerializableResource.new(
+        @products, each_serializer: ProductSerializer
+      )
+    }
+  end
+
+  def show
+    @product = Product.find(params[:id])
```
Comment (nghinn1012) [inline app/controllers/api/v1/admin/products_controller.rb]: find_by, nếu không thấy thì trả mảng rỗng chứ ntn thì lỗi

File: app/controllers/api/v1/admin/products_controller.rb [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    else
+      render json: {errors: @product.errors.full_messages},
+             status: :unprocessable_entity
+    end
+  end
+
+  def update
+    @product = Product.find(params[:id])
+
+    if @product.update(product_params)
```
Comment (nghinn1012) [inline app/controllers/api/v1/admin/products_controller.rb]: tương tự, xử lý không tìm thấy

File: app/controllers/api/v1/admin/products_controller.rb [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+      render json: @product, serializer: ProductSerializer
+    else
+      render json: {errors: @product.errors.full_messages},
+             status: :unprocessable_entity
+    end
+  end
+
+  def destroy
+    @product = Product.find(params[:id])
```
Comment (nghinn1012) [inline app/controllers/api/v1/admin/products_controller.rb]: đây nữa

File: app/controllers/api/v1/products_controller.rb [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    ).as_json
+
+    render json: {
+      status: true,
+      data: serialized_products[:data]
+    }
+  end
+
+  def show
+    @product = Product.find(params[:id])
```
Comment (nghinn1012) [inline app/controllers/api/v1/products_controller.rb]: như trên, dùng find_by

File: app/views/devise/passwords/edit.html.erb [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    <%= f.password_field :password, autofocus: true, autocomplete: "new-password" %>
+  </div>
+
+  <div class="field">
+    <%= f.label :password_confirmation, "Confirm new password" %><br />
+    <%= f.password_field :password_confirmation, autocomplete: "new-password" %>
+  </div>
+
+  <div class="actions">
+    <%= f.submit "Change my password" %>
```
Comment (nghinn1012) [inline app/views/devise/passwords/edit.html.erb]: i18n nè

File: app/views/devise/passwords/new.html.erb [hunk 6]
```diff
@@ -0,0 +1,16 @@
+<h2>Forgot your password?</h2>
+
+<%= form_for(resource, as: resource_name, url: password_path(resource_name), html: { method: :post }) do |f| %>
+  <%= render "devise/shared/error_messages", resource: resource %>
+
```
Comment (nghinn1012) [inline app/views/devise/passwords/new.html.erb]: i18n

**Comment by ThaiKhacDuoc** [comment]
ready

**Comment by ThaiKhacDuoc** [comment]
ready

**Review by ghost — APPROVED** [review]

**Pull request merged**